api/turing/api/ensemblers_api.go

```diff
@@ -109,18 +111,172 @@ func (c EnsemblersController) UpdateEnsembler(
 			"Ensembler type cannot be changed after creation")
 	}
 
+	pyFuncEnsembler, isPyfunc := ensembler.(*models.PyFuncEnsembler)
+
+	var oldPyFuncEnsembler models.PyFuncEnsembler
+	if isPyfunc {
+		oldPyFuncEnsembler = models.PyFuncEnsembler{
+			RunID:       pyFuncEnsembler.RunID,
+			ArtifactURI: pyFuncEnsembler.ArtifactURI,
+		}
+	}
+
 	if err = ensembler.Patch(request.EnsemblerLike); err != nil {
 		return BadRequest("invalid ensembler configuration", err.Error())
 	}
-
 	ensembler, err = c.EnsemblersService.Save(ensembler)
 	if err != nil {
+		// Delete If Only RunID is changed, if only name is changed, doesn't need to delete the mlflow run
+		if isPyfunc && oldPyFuncEnsembler.RunID != pyFuncEnsembler.RunID {
+			err = c.MlflowService.DeleteRun(context.Background(), pyFuncEnsembler.RunID, pyFuncEnsembler.ArtifactURI, true)
+			if err != nil {
+				return InternalServerError("failed to update an ensembler", "cleanup process failed")
+			}
+		}
 		return InternalServerError("failed to update an ensembler", err.Error())
 	}
 
+	// Delete If Only RunID is changed, if only name is changed, doesn't need to delete the mlflow run
+	if isPyfunc && oldPyFuncEnsembler.RunID != pyFuncEnsembler.RunID {
+		_ = c.MlflowService.DeleteRun(context.Background(), oldPyFuncEnsembler.RunID, oldPyFuncEnsembler.ArtifactURI, true)
+	}
+
 	return Ok(ensembler)
 }
 
+func (c EnsemblersController) DeleteEnsembler(
+	_ *http.Request,
+	vars RequestVars,
+	_ interface{},
+) *Response {
+	options := EnsemblersPathOptions{}
+
+	if err := c.ParseVars(&options, vars); err != nil {
+		return BadRequest("failed to fetch ensembler",
+			fmt.Sprintf("failed to parse query string: %s", err))
+	}
+
+	ensembler, err := c.EnsemblersService.FindByID(
+		*options.EnsemblerID,
+		service.EnsemblersFindByIDOptions{
+			ProjectID: options.ProjectID,
+		})
+
+	if err != nil {
+		return NotFound("ensembler not found", err.Error())
+	}
+
+	// CHECK IF STATUS ROUTER IS DEPLOYED
+	routerVersionStatusActive := []models.RouterVersionStatus{
+		models.RouterVersionStatusDeployed,
+		models.RouterVersionStatusPending,
+	}
+	activeOption := service.RouterVersionByEnsemblerListOptions{
+		ProjectID:   options.ProjectID,
+		EnsemblerID: options.EnsemblerID,
+		Statuses:    routerVersionStatusActive,
+	}
+	activeRouter, err := c.RouterVersionsService.FindRouterVersionsByEnsembler(activeOption)
+	if err != nil {
+		return InternalServerError("failed to delete an ensembler", err.Error())
+	}
+	if len(activeRouter) >= 1 {
```

Comment:
Just to clarify, this new API will be used only in frontend right? since the API of router version deletion already handle this, so we dont need to check in the backend again.

Reply:
My bad, I forgot that it's possible to have router versions that are not active but are 'current'. Thanks a lot for clarifying this @zenovore!